src/Orchard/Tasks/Locking/Services/DistributedLockService.cs

```
@@ -0,0 +1,153 @@
+        private readonly IClock _clock;
+        private readonly object _semaphore = new object();
+
+        public DistributedLockService(IMachineNameProvider machineNameProvider, ILifetimeScope lifetimeScope, IClock clock) {
+            _machineNameProvider = machineNameProvider;
+            _lifetimeScope = lifetimeScope;
+            _clock = clock;
+            Logger = NullLogger.Instance;
+        }
+
+        public ILogger Logger { get; set; }
+
+        public bool TryAcquireLock(string name, TimeSpan maxLifetime, TimeSpan timeout, out IDistributedLock @lock) {
+            lock (_semaphore) {
+                @lock = default(IDistributedLock);
+
+                try {
+                    var waitedTime = TimeSpan.Zero;
+                    var waitTime = TimeSpan.FromMilliseconds(timeout.TotalMilliseconds / 10);
+                    bool acquired;
+
+                    while (!(acquired = TryAcquireLockRecord(name, maxLifetime, out @lock)) && waitedTime < timeout) {
+                        Task.Delay(timeout).ContinueWith(t => {
+                            waitedTime += waitTime;
+                        }).Wait();
+                    }
+
+                    if (acquired) {
+                        Logger.Debug("Successfully acquired a lock named {0}.", name);
+                        return true;
+                    }
+                }
+                catch (Exception ex) {
+                    Logger.Error(ex, "Error while trying to acquire a lock named {0}.", name);
+                    throw;
+                }
+
+                Logger.Debug("Could not acquire a lock named {0}.", name);
+                return false;
+            }
+        }
+
+        public IDistributedLock AcquireLock(string name, TimeSpan maxLifetime, TimeSpan timeout) {
+            IDistributedLock lockResult;
+            return TryAcquireLock(name, maxLifetime, timeout, out lockResult) ? lockResult : null;
```

This should throw a `TimeoutException` if the lock could not be acquired, rather than returning `null`.